corona-warn-app/cwa-app-android
# Better handling of issues caused by time traveling (RiskResults) (DEV)

Change primary key for risk result database to a monotonic ID.

The `RiskLevelTask` calculates a `RiskResult`, that's stored in a database.
The UI uses the "latest result" and "latest successful result" when displaying the risk/tracing card on the home screen.
If the device travels into the future, calculates a result, and then returns back to "our" time, then the risk card always has a "newer" result until we have caught up to the future date at which the result was calculated.

This is also likely not the cause for #2001, but looking deeper into that issue caused me to think it's nice to have this additional resilience against time-traveling issues. This also makes it easier for developers/users/testers to debug issues by changing the devices clock, this was previously possible before the move to ENFv2.

The PR migrates the database by creating a new table that has the autoincrementing field `monotonicId` and then moving the old data into it. The SQL queries are adjusted to sort my `monotonicId` instead of `calculatedAt`.

## Testing
* Use the app without this PR, fake a few different risk results through the test menu
* Update to this PR's version
* Everything should look the same

* Use this app without this PR, time travel and generate a risk result in the future
* Return to normal time, notice that any risk result changes (e.g through the test menu), will not replace the result (card state) from the future
* Upgrade to this PR's version, the risk result card on the home screen should now show the last calculated result (which is from your time now, and not the one from time traveling)
